content/en/blog/2024/building-trust-into-os-images-for-coco.md

```diff
@@ -0,0 +1,191 @@
+
+```bash
+$ dd if=/dev/zero of=rootfs.raw bs=1M count=100
+$ DEVICE="$(sudo losetup --show -f rootfs.raw)"
+$ sudo cfdisk "$DEVICE"
+# create 1 partition
+$ sudo mkfs.ext4 "$DEVICE"
+...
+$ sudo mount "$DEVICE" /mnt
+$ echo "CoCo" | sudo tee /mnt/coco
+CoCo
+$ sudo umount /mnt
+$ sudo veritysetup format "$DEVICE" ./hash-tree
+VERITY header information for ./hash-tree
+UUID:                   91bbc990-f0df-48c0-b8f0-1b996cf0c3cf
+Hash type:              1
+Data blocks:            25600
+Data block size:        4096
+Hash block size:        4096
+Hash algorithm:         sha256
+Salt:                   cef7ea72e3487f4f8d26df8731df561f64e03236fa494dc0ae87fe0f07a4825b
+Root hash:              ad86ff8492be2ee204cb54d70c84412c2dc89cefd34e263184f4e00295a412f3
+$ export ROOT_HASH=ad86ff8492be2ee204cb54d70c84412c2dc89cefd34e263184f4e00295a412f3
+```
+
+### Corrupting the image
+
+Now we toggle a bit on the raw image (`CoCo` => `DoCo` in `/coco`). If the image is attached as a block device via dm-verity, there will be IO errors and respective entries in the kernel log, once we attempt to read the file.
+
+```bash
+$ hexdump -C rootfs.raw | grep CoCo
+06000000  43 6f 43 6f 0a 00 00 00  00 00 00 00 00 00 00 00  |CoCo............|
+$ printf '\x44' | dd of=rootfs.raw bs=1 seek="$((16#06000000))" count=1 conv=notrunc
+$ hexdump -C rootfs.raw | grep DoCo
+06000000  44 6f 43 6f 0a 00 00 00  00 00 00 00 00 00 00 00  |DoCo............|
+$ sudo veritysetup open "$DEVICE" rootfs ./hash-tree "$ROOT_HASH"
+$ sudo mount /dev/mapper/rootfs /mnt
+$ cat /mnt/coco
+cat: /mnt/coco: Input/output error
+$ dmesg | tail -1
+[194754.361797] device-mapper: verity: 7:0: data block 24576 is corrupted
+```
+
+## vTPM
+
+![](/img/os-images-vtpm.png)
+
+There are setups in which a launch measurement of the TEEs will not cover the kernel and/or initrd. An example of such a TEE is Azure's Confidential VM offering (L1 VMs provided by hypervisor running on a physical host). Those CVMs can host Confidential Containers in a [CoCo Peerpod setup](https://confidentialcontainers.org/docs/cloud-api-adaptor/). The hardware evidence, which is attesting encrypted RAM and CPU registers is exclusively fetched during an [early boot phase](https://techcommunity.microsoft.com/t5/windows-os-platform-blog/confidential-vms-on-azure/ba-p/3836282). Only in later stages the kernel and initrd are loaded from an OS image and hence the launch measurement will not cover the CoCo infrastructure components yet. To still be able to provide an integrity guarantee such a TEE can defer measurements of the later boot stages to a virtual TPM device (vTPM) that is being created on the confidential guest in [privilege level](https://www.amd.com/content/dam/amd/en/documents/epyc-business-docs/white-papers/SEV-SNP-strengthening-vm-isolation-with-integrity-protection-and-more.pdf) that is isolated during the lifecycle of the CVM and cryptographically linked to the hardware evidence.
```

Comment:
Maybe beyond the scope of the article, but it might be worth noting the difference between traditional vTPMs and confidential ones.